### DISCUSSION TITLE
Using getServerSideProps inside a component

### DISCUSSION THREAD

**Discussion by mafshin** [vercel/next.js#11590] [Help]
I want to dynamically get data based on the query parameter. 

Using `getServerSideProps` in pages work, but how can I do the same (getting data server side only) in components?

**Comment by lkbr** [comment]
Fetch on the page and pass props down (or use React context). 

I'm not on the next team, so this view is observation only: there is tremendous value in keeping the get* functions exclusively and predictably in pages that unlocks many of the recent and upcoming Next features.

**Reply by mafshin** [reply]
Thank you @lkbr, already tried that and it works but I have to duplicate the data fetching logic in each page.

**Reply by lkbr** [reply]
Yes, thats inconvenient. You could try extracting the data fetching logic into a function and then use that function in each pages `getServerSideProps`.

**Comment by padcom** [comment]
As with everything-React things aren't as simple as they seem. Sure, Next.js does allow you to do server-side props fetching for page components (it's still just a React component but is a "Page" because it also exports `getServerSideProps()`) but it doesn't do that for any other components. Yet again, we find ourselves in a situation where a component is still a component unless it isn't just a component, then it is not a component but a page. I honestly find that not only difficult to understand why but also very frustrating. The "element of the least surprise" rule is harshly violated here...

But what you can do is to be just a bit smart(er) about the code you're writing and instead forcibly shoving everything into the page's `getServerSideProps()` function instead export one from the component you want and then in the Page component import it like so:

```
import SomeComponent, { getServerSideProps as someComponentGetServerSideProps } from '../components/SomeComponent'
```

and then call it from the page's `getServerSideProps()`:

```
export async function getServerSideProps(context) {
  return {
    props: {
      data: someComponentGetServerSideProps(context)
    }
  }
}
```

Something like that will make your frontend code co-exist with the server-side data provider, which will then mean if you need to remove your component at a later time, you'll have (almost) everything in one place (the `SomeComponent.tsx`). The one thing that will still need to be updated is the Page component, where you will still need to remove the call to `someComponentGetServerSideProps` manually from `getServerSideProps()`. But at least if the `someComponentGetServerSideProps()` implementation changes without modifying the interface, then you won't need to touch your Page.

Of course, don't name the export the way I did it. It's just an example!

Hope it helps.
